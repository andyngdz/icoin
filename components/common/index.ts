export { ICoinLogo } from 'components/common/Logo'
export { Route } from 'components/common/Route'
