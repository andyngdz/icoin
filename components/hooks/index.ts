export * from './useChartData'
export * from './useGlobal'
export * from './useLivePrice'
export * from './useLivePrices'
export * from './usePagination'
export * from './useTheme'
export * from './useTime'
