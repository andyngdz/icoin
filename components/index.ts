export * from './asset-detail/AssetDetail'
export * from './asset-detail/AssetDetailContent'
export * from './asset-detail/AssetHighLow'
export * from './asset-detail/AssetSummary'
export * from './assets/AssetItem'
export * from './assets/AssetNameAndDescription'
export * from './assets/Assets'
export * from './assets/AssetsTable'
export * from './common/container-wrapper/ContainerWrapper'
export * from './common/logo/Logo'
export * from './common/pagination/Pagination'
export * from './common/paper-wrapper/PaperWrapper'
export * from './common/route/Route'
export * from './common/value-color/ValueColor'
export * from './disclaimer/Disclaimer'
export * from './exchanges/ExchangeContent'
export * from './exchanges/ExchangeItem'
export * from './exchanges/ExchangesAll'
export * from './exchanges/ExchangesAsset'
export * from './exchanges/ExchangeStatus'
export * from './header/Header'
export * from './header/Menu'
export * from './hooks'
export * from './line-chart/LineChart'
export * from './live-prices/LivePrices'
export * from './search/Search'
export * from './symbol-icon/SymbolIcon'
export * from './time-selection/TimeSelection'
export * from './top-bar/TopBar'
export * from './top-bar/TopBarContent'
