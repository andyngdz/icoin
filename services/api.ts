import { Format } from 'services'
import { IAPI, IAssetResponse } from 'types'
import { PER_PAGE } from 'data'
import axios from 'axios'

const CoingeckoAPI = axios.create({
  baseURL: process.env.NEXT_PUBLIC_COINGECKO_API
})

const CoinCapAPI = axios.create({
  baseURL: process.env.NEXT_PUBLIC_COINCAP_API
})

const API: IAPI = {
  getGlobals: () => CoingeckoAPI.get('/global'),

  getAssets: async (
    params = {
      limit: PER_PAGE
    }
  ) => {
    const response = await CoinCapAPI.get<IAssetResponse>('/assets', {
      params
    })
    const { data } = response.data

    response.data.data = Format.toNumber(data, [
      'marketCapUsd',
      'priceUsd',
      'volumeUsd24Hr',
      'changePercent24Hr',
      'supply',
      'maxSupply',
      'vwap24Hr'
    ])

    return response
  }
}

export { API }
