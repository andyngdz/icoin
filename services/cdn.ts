const CDN = {
  getIcon: (id: string) =>
    `https://cdn.jsdelivr.net/gh/atomiclabs/cryptocurrency-icons@9ab8d6934b83a4aa8ae5e8711609a70ca0ab1b2b/svg/color/${id}.svg`,

  defaultIcon: `https://cdn.jsdelivr.net/gh/atomiclabs/cryptocurrency-icons@9ab8d6934b83a4aa8ae5e8711609a70ca0ab1b2b/svg/color/generic.svg`,
}

export { CDN }
