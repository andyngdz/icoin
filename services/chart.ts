/**
 * Author: Andy D. Ng <itc.anhduy@gmail.com>
 * Chart.js contains the useful function to manage a Chart
 */

import { ChartData, ChartOptions, TimeScale } from 'chart.js'
import { IAssetHistory, ICalculateInterval, TTime } from 'types'
import { merge } from 'lodash'
import { sub } from 'date-fns'
import { TIME_TO_INTERVAL } from 'data'
import ChartJS from 'chart.js'

/**
 * @description
 * We are using Roboto font and
 * Set it here for the ChartJS as well
 */
ChartJS.defaults.global.defaultFontFamily = "'Inter', sans-serif"

const Chart = {
  /**
   * @description Calculate interval post data by id, time, start and end date
   *
   * @param id The id of coin such as: bitcoin, ethereum
   *
   * @param time The time from the TimeSelection component - (Check TTime in chart.d.ts)
   *
   * @returns `ICalculateInterval`
   *
   * @example
   * ```
   *    const { time, onTimeChange } = useTime()
   *     const [getIntervals, { data }] = useLazyQuery<IAssetHistories>(COIN_CHART)
   *
   *     useEffect(() => {
   *       const interval = Chart.calculateInterval(id, time)
   *       getIntervals({ variables: interval })
   *     }, [time])
   *```
   */
  calculateInterval: (id: string, time: TTime): ICalculateInterval => {
    let subtractDate: Date
    const end = Date.now()

    switch (time) {
      case '1D':
        subtractDate = sub(end, { days: 1 })
        break
      case '1W':
        subtractDate = sub(end, { weeks: 1 })
        break
      case '1M':
        subtractDate = sub(end, { months: 1 })
        break
      case '3M':
        subtractDate = sub(end, { months: 3 })
        break
      case '6M':
        subtractDate = sub(end, { months: 6 })
        break
      case '1Y':
      case 'ALL':
        subtractDate = sub(end, { years: 1 })
        break
    }

    return {
      end,
      id,
      interval: TIME_TO_INTERVAL[time],
      start: subtractDate.valueOf()
    }
  },

  /**
   * @description Create time scale with unit for Line Chart
   *
   * @param time The time from the TimeSelection component - (Check TTime in chart.d.ts)
   *
   * @returns `TimeScale`
   */
  createTimeScale: (time: TTime): TimeScale => {
    const timeScale: TimeScale = {}

    switch (time) {
      case '1D':
        timeScale.unit = 'hour'
        break
      case '1W':
        timeScale.unit = 'day'
        break
      case '1M':
      case '3M':
        timeScale.unit = 'week'
        break
      case '6M':
      case '1Y':
      case 'ALL':
        timeScale.unit = 'month'
        break
    }

    return timeScale
  },

  /**
   * @description Create a new chart
   *
   * @param ctx The canvas rendering context
   *
   * @param assetHistories Contains the list history data such as: timestamp, priceUsd, etc...
   *
   * @param options The chart options, it will be merged with the default options
   *
   * @returns `ChartJS`
   */
  createNewChart: (
    canvas: HTMLCanvasElement,
    time: TTime,
    chartData: ChartData,
    options?: ChartOptions
  ): ChartJS => {
    const ctx = canvas.getContext('2d')

    return new ChartJS(ctx, {
      type: 'line',
      data: chartData,
      options: merge(
        {
          tooltips: {
            position: 'nearest'
          },
          legend: {
            display: false
          },
          scales: {
            xAxes: [
              {
                type: 'time',
                time: Chart.createTimeScale(time),
                gridLines: {
                  display: false
                }
              }
            ],
            yAxes: [
              {
                position: 'right',
                gridLines: {
                  drawBorder: false
                }
              }
            ]
          }
        },
        options
      )
    })
  },

  /**
   * @description Create chart data
   *
   * @param assetHistories Contains the list history data such as: timestamp, priceUsd, etc...
   *
   * @returns `ChartData`
   */
  createChartData: (assetHistories: Array<IAssetHistory>): ChartData => {
    const data = assetHistories.map(({ timestamp, priceUsd }) => ({
      t: timestamp,
      y: priceUsd
    }))

    const chartData: ChartData = {
      datasets: [
        {
          data
        }
      ]
    }

    return chartData
  }
}

export { Chart }
