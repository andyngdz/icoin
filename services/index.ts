export * from './paginate'
export * from './chart'
export * from './format'
export * from './api'
export * from './cdn'
export * from './test-utils'
export * from './price-direction'
export * from './price'
export * from './routes'
