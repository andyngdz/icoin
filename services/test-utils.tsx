import { Provider } from 'react-redux'
import { render } from '@testing-library/react'
import { ThemeProvider } from '@material-ui/core'
import React from 'react'
import Store from 'states'
import Theme from 'styles'

const Providers = ({ children }) => {
  return (
    <Provider store={Store}>
      <ThemeProvider theme={Theme}>{children}</ThemeProvider>
    </Provider>
  )
}

const customRender = (ui, options = {}) =>
  render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'

export { customRender as render }
