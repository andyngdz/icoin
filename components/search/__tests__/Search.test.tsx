import { customRender } from 'services/test-utils'
import { Search } from 'components'

describe('Search', () => {
  it('Should render it correctly', async () => {
    const renderResult = customRender(<Search />)
    expect(renderResult).toMatchSnapshot()
  })
})
