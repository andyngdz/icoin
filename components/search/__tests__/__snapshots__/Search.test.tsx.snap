// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Should render it correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section>
        <div
          class="MuiPaper-root Search-search-1 MuiPaper-elevation6"
        >
          <div
            class="MuiContainer-root ContainerWrapper-container-4 MuiContainer-maxWidthLg"
          >
            <div
              class="MuiFormControl-root MuiTextField-root Search-input-2 MuiFormControl-fullWidth"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                  placeholder="Search for cryptocurrencies"
                  type="text"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
                  style="padding-left: 8px;"
                >
                  <legend
                    class="PrivateNotchedOutline-legend-6"
                    style="width: 0.01px;"
                  >
                    <span>
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="Search-overlay-3"
          style="opacity: 0; visibility: hidden;"
        />
      </section>
    </div>
  </body>,
  "container": <div>
    <section>
      <div
        class="MuiPaper-root Search-search-1 MuiPaper-elevation6"
      >
        <div
          class="MuiContainer-root ContainerWrapper-container-4 MuiContainer-maxWidthLg"
        >
          <div
            class="MuiFormControl-root MuiTextField-root Search-input-2 MuiFormControl-fullWidth"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                placeholder="Search for cryptocurrencies"
                type="text"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-6"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="Search-overlay-3"
        style="opacity: 0; visibility: hidden;"
      />
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
